### ISSUE TITLE
Cannot cast java AtomicLong exception and custom Concurrent connections is leaky over time and 

### ISSUE THREAD

**Issue by AkashKumarAM21** [reactor/reactor-netty#2992]
We have two issue, which we think could be related.
* Netty infrequently throws - 
```java
encountered due to root cause: java.lang.ClassCastException: class reactor.netty.http.server.HttpServerOperations cannot be cast to class java.util.concurrent.atomic.AtomicLong (reactor.netty.http.server.HttpServerOperations is in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @59e84876; java.util.concurrent.atomic.AtomicLong is in module java.base of loader 'bootstrap')
com.xxxxx.xxx.xxx: Unrecognized Exception: Unexpected exception from Netty
      at com.xxxxx.xxxxx(xxxxx.java:196) ~[classes!/:?]
      at com.xxxxx.xxxxx(xxxxx.java:101) ~[classes!/:?]
      at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
      at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: java.lang.RuntimeException: Unexpected exception from Netty
      at com.xxxxx.xxxxx(xxxxx.java:87) ~[classes!/:?]
      ... 6 more
Caused by: java.lang.ClassCastException: class reactor.netty.http.server.HttpServerOperations cannot be cast to class java.util.concurrent.atomic.AtomicLong (reactor.netty.http.server.HttpServerOperations is in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @59e84876; java.util.concurrent.atomic.AtomicLong is in module java.base of loader 'bootstrap')
      at reactor.netty.http.server.HttpServerOperations.requestsCounter(HttpServerOperations.java:756) ~[reactor-netty-http-1.0.27.jar!/:1.0.27]
      at reactor.netty....<TRUNCATED>...5) ~[reactor-netty-http-1.0.27.jar!/:1.0.27]
      at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:709) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:792) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:702) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at reactor.netty.http.server.AbstractHttpServerMetricsHandler.write(AbstractHttpServerMetricsHandler.java:154) ~[reactor-netty-http-1.0.27.jar!/:1.0.27]
      at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:709) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:792) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:702) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:110) ~[Netty4-2-4.1.jar!/:4.1.78.Final]
      at io.netty.handler.codec.MessageToMessageCodec.write(MessageToMessageCodec.java:116) ~[Netty4-2-4.1.jar!/:4.1.78.Final]
      at reactor.netty.http.server.SimpleCompressionHandler.write(SimpleCompressionHandler.java:45) ~[reactor-netty-http-1.0.27.jar!/:1.0.27]
      at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:764) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1071) ~[Netty4-22-4.1.jar!/:4.1.78.Final]
      at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:174) ~[Netty4-14-4.1.jar!/:4.1.78.Final]
      at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:167) ~[Netty4-14-4.1.jar!/:4.1.78.Final]
      at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[Netty4-14-4.1.jar!/:4.1.78.Final]
      at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:391) ~[Netty4-23-4.1.jar!/:4.1.78.Final]
      at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[Netty4-14-4.1.jar!/:4.1.78.Final]
      at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[Netty4-14-4.1.jar!/:4.1.78.Final]
      at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[Netty4-14-4.1.jar!/:4.1.78.Final]
      ... 1 more
```

* We emit our own simple custom concurrent connections metric, that metric over time keeps on growing, only decreases when we reboot the server. 

```java
private final AtomicInteger concurrentConnections = new AtomicInteger();
 
public Publisher<Void> handle(HttpServerRequest httpServerRequest, HttpServerResponse httpServerResponse) {
        emitMetric(concurrentConnections.incrementAndGet());
         
        try {
            SubscriberInputStream subscriberInputStream = new SubscriberInputStream();
            try (final CloseableThreadContext.Instance ignored = CloseableThreadContext.put("RequestId", requestId)) {
                try {
                    return rateLimiter.executeSupplier(() -> {
                        return handleRequest();
                    });
                } catch (RequestNotPermitted e) {
                    return handleFailure();
                }
            }
        } catch (Throwable e) {
            return handleFailure();
        }
    }
 
    private Publisher<Void> handleRequest() {
        CompletableFuture<Void> result = CompletableFuture.runAsync(() -> {
            try () {
                // action 
            } finally {
                concurrentConnections.decrementAndGet();
            }
        }, executor);
        return Mono.fromCompletionStage(result);
    }
 
    private Publisher<Void> handleFailure() {
        concurrentConnections.decrementAndGet();
    }
```

## Expected Behavior
* We are expecting the `concurrentConnections` to decrement when we complete handling the request, either successfully or due to error and we are not sure why Netty throws casting exception. 

## Your Environment

* Reactor version(s) used: io.projectreactor_reactor-core-3.4.24
* Other relevant libraries versions (eg. `netty`, ...): Netty 4.1
* JVM version (`java -version`): JDK17
* OS and version (eg. `uname -a`): AL2_x86_64

**Comment by violetagg** [comment]
@AkashKumarAM21 Another thing that you might try to check is the change of the class loaders described here https://github.com/spring-projects/spring-boot/issues/19427

**Comment by AkashKumarAM21** [comment]
> @AkashKumarAM21 Can you check for two versions of Reactor Netty in your dependencies?

We can't have duplicate versions of same dependency and we fail if there duplicate class paths too, our build tools enforce it! I'll double check it though.

**Comment by AkashKumarAM21** [comment]
This is the latest stack trace, after we upgraded the dependencies. 

```java
encountered due to root cause: java.lang.ClassCastException: class reactor.netty.http.server.HttpServerOperations cannot be cast to class java.util.concurrent.atomic.AtomicLong (reactor.netty.http.server.HttpServerOperations is in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @39fb3ab6; java.util.concurrent.atomic.AtomicLong is in module java.base of loader 'bootstrap')
	at com.xxxx.xxxxx(xxxx.java:106) ~[classes!/:?]
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: java.lang.RuntimeException: Unexpected exception from Netty
	at com.xxxx.xxxx(xxxx.java:89) ~[classes!/:?]
	at java.io.OutputStream.write(OutputStream.java:162) ~[?:?]
	at java.io.InputStream.transferTo(InputStream.java:783) ~[?:?]
	at com.xxxx.xxxx(xxxx.java:252) ~[classes!/:?]
	... 6 more
	Suppressed: java.lang.RuntimeException: Unexpected exception from Netty
		at com.xxxx.xxxx(xxxxxxx.java:89) ~[classes!/:?]
		at com.xxxx.xxxx(xxxxxxx.java:66) ~[classes!/:?]
		at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
		at java.lang.Thread.run(Thread.java:840) [?:?]
	Caused by: java.lang.ClassCastException: class reactor.netty.http.server.HttpServerOperations cannot be cast to class java.util.concurrent.atomic.AtomicLong (reactor.netty.http.server.HttpServerOperations is in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @39fb3ab6; java.util.concurrent.atomic.AtomicLong is in module java.base of loader 'bootstrap')
		at reactor.netty.http.server.HttpServerOperations.requestsCounter(HttpServerOperations.java:788) ~[reactor-netty-http-1.0.36.jar!/:1.0.36]
		at reactor.netty.http.server.HttpTrafficHandler.write(HttpTrafficHandler.java:307) ~[reactor-netty-http-1.0.36.jar!/:1.0.36]
		at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at reactor.netty.http.server.AbstractHttpServerMetricsHandler.write(AbstractHttpServerMetricsHandler.java:154) ~[reactor-netty-http-1.0.36.jar!/:1.0.36]
		at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:113) ~[Netty4-2-4.1.jar!/:4.1.100.Final]
		at io.netty.handler.codec.MessageToMessageCodec.write(MessageToMessageCodec.java:116) ~[Netty4-2-4.1.jar!/:4.1.100.Final]
		at reactor.netty.http.server.SimpleCompressionHandler.write(SimpleCompressionHandler.java:50) ~[reactor-netty-http-1.0.36.jar!/:1.0.36]
		at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:940) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1247) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
		at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
		at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
		at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:416) ~[Netty4-24-4.1.jar!/:4.1.100.Final]
		at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
		... 1 more
Caused by: java.lang.ClassCastException: class reactor.netty.http.server.HttpServerOperations cannot be cast to class java.util.concurrent.atomic.AtomicLong (reactor.netty.http.server.HttpServerOperations is in unnamed module of loader org.springframework.boot.loader.LaunchedURLClassLoader @39fb3ab6; java.util.concurrent.atomic.AtomicLong is in module java.base of loader 'bootstrap')
	at reactor.netty.http.server.HttpServerOperations.requestsCounter(HttpServerOperations.java:788) ~[reactor-netty-http-1.0.36.jar!/:1.0.36]
	at reactor.netty.http.server.HttpTrafficHandler.write(HttpTrafficHandler.java:307) ~[reactor-netty-http-1.0.36.jar!/:1.0.36]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at reactor.netty.http.server.AbstractHttpServerMetricsHandler.write(AbstractHttpServerMetricsHandler.java:154) ~[reactor-netty-http-1.0.36.jar!/:1.0.36]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:863) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:968) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:856) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:113) ~[Netty4-2-4.1.jar!/:4.1.100.Final]
	at io.netty.handler.codec.MessageToMessageCodec.write(MessageToMessageCodec.java:116) ~[Netty4-2-4.1.jar!/:4.1.100.Final]
	at reactor.netty.http.server.SimpleCompressionHandler.write(SimpleCompressionHandler.java:50) ~[reactor-netty-http-1.0.36.jar!/:1.0.36]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:879) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:940) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.AbstractChannelHandlerContext$WriteTask.run(AbstractChannelHandlerContext.java:1247) ~[Netty4-23-4.1.jar!/:4.1.100.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:416) ~[Netty4-24-4.1.jar!/:4.1.100.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[Netty4-14-4.1.jar!/:4.1.100.Final]
	... 1 more
	```

**Comment by AkashKumarAM21** [comment]
> @AkashKumarAM21 Another thing that you might try to check is the change of the class loaders described here [spring-projects/spring-boot#19427](https://github.com/spring-projects/spring-boot/issues/19427)

We already provide custom Executor for our `runAsync` call, so this is not the issue ?

**Comment by violetagg** [comment]
@AkashKumarAM21 Can you provide a reproducible example so that we can investigate the issue?

**Comment by AkashKumarAM21** [comment]
> @AkashKumarAM21 Can you provide a reproducible example so that we can investigate the issue?

Thats where we came to a blocker, this exception is infrequently (like 1 request fails for every 200,000 requests) and we are unable to produce it. We tried upgrading our dependencies to check if thats going to fix the issue, not luck yet.

**Comment by violetagg** [comment]
@AkashKumarAM21 It is hard for us to proceed without a clear scenario/reproducible example.

**Comment by AkashKumarAM21** [comment]
> @AkashKumarAM21 It is hard for us to proceed without a clear scenario/reproducible example.

Yup, i understand that! we don't have any reproduction steps :(

Is there any feature/property that we can enable to get more information ? It would be helpfully if there is any logging or instrumentation we can turn on to get more details in this scenario.

**Comment by violetagg** [comment]
@AkashKumarAM21 Reactor Netty never changes the class loaders, we don't have any logging about this.

**Comment by violetagg** [comment]
@AkashKumarAM21 I'm closing this as I cannot reproduce the issue. We can reopen it, if you are able to provide reproducible example.

**Comment by alexexternaltest35-ctrl** [comment]
Hi @violetagg 

We are seeing this issue chronically (but infrequently). This is still happening like 1 in every hundred of thousands of requests during high concurrent traffic, so unfortunately we cannot get a working reproduction example (though sometimes I can get it to trigger by just spinning up a server locally and spamming it with concurrent keep-alive requests).

We are using JDK21 with reactor-netty-http 1.2.15. Response bodies are written from a virtual thread through a reactive streams Publisher. The virtual thread produces bytes, reactor-netty subscribes and writes them to the channel. When the publisher signals onComplete(), reactor-netty writes the final LastHttpContent frame and begins connection cleanup. We use HTTP/1.1 keep-alive with no pipelining.

I was kind of able to reproduce the exception very infrequently locally and enabled debug logs for Netty. I think this is some sort of race condition. What we see is this:

  Crashing channel: next request processed before terminate:
```
  25 Mar 2026 21:19:30,517 [DEBUG] (virtual-thread)      HttpServerOperations: [6868f17c-11, L:/10.0.0.1:8080 - R:/10.0.0.2:39860] Last HTTP response frame
  25 Mar 2026 21:19:30,517 [DEBUG] (reactor-http-epoll-7) HttpServerOperations: [6868f17c-11, L:/10.0.0.1:8080 - R:/10.0.0.2:39860] Decreasing pending responses count: 0
  25 Mar 2026 21:19:30,518 [DEBUG] (reactor-http-epoll-7) HttpServerOperations: [6868f17c-11, L:/10.0.0.1:8080 - R:/10.0.0.2:39860] Increasing pending responses count: 1
  25 Mar 2026 21:19:30,519 [DEBUG] (reactor-http-epoll-7) HttpServer:           [6868f17c, L:/10.0.0.1:8080 - R:/10.0.0.2:39860] Handler is being applied: DefaultHttpServerRoutes@3098c25
  25 Mar 2026 21:19:30,519 [DEBUG] (reactor-http-epoll-7) HttpServerOperations: [6868f17c, L:/10.0.0.1:8080 - R:/10.0.0.2:39860] Last HTTP packet was sent, terminating the channel
```
  
  Safe channel at the same instant: terminate completes first:
```
  25 Mar 2026 21:19:30,517 [DEBUG] (virtual-thread)      HttpServerOperations: [7181b496-12, L:/10.0.0.1:8080 - R:/10.0.0.2:39872] Last HTTP response frame
  25 Mar 2026 21:19:30,518 [DEBUG] (reactor-http-epoll-8) HttpServerOperations: [7181b496-12, L:/10.0.0.1:8080 - R:/10.0.0.2:39872] Decreasing pending responses count: 0
  25 Mar 2026 21:19:30,518 [DEBUG] (reactor-http-epoll-8) HttpServerOperations: [7181b496-12, L:/10.0.0.1:8080 - R:/10.0.0.2:39872] Last HTTP packet was sent, terminating the channel
  25 Mar 2026 21:19:30,518 [DEBUG] (reactor-http-epoll-8) HttpServerOperations: [7181b496, L:/10.0.0.1:8080 - R:/10.0.0.2:39872] Increasing pending responses count: 1
  25 Mar 2026 21:19:30,519 [DEBUG] (reactor-http-epoll-8) HttpServer:           [7181b496-13, L:/10.0.0.1:8080 - R:/10.0.0.2:39872] Handler is being applied: DefaultHttpServerRoutes@3098c25
```
  
  The difference: on the crashing channel, "Increasing pending responses count: 1" (next request's channelRead) fires before "Last HTTP packet was sent, terminating the channel" (terminate()/rebind()). On the safe
  channel, terminate completes first.

*Disclaimer*: This is mostly Claude looking through the netty source code:

What we think is happening:
  
  The race is between ctx.read() and the flush-complete callback that runs terminate()/rebind(). In HttpTrafficHandler.handleLastHttpContent(), ctx.read() is called immediately after ctx.write(), before the flush
  completes. On a keep-alive connection where the next request's bytes are already in the socket receive buffer, channelRead for the next request can fire before rebind() has restored the CONNECTION attribute back to SimpleConnection.

We saw this exception more frequently when upgrading from 1.0.40 to 1.2.15. According to Claude, 
"in 1.0.40 HttpTrafficHandler itself was the flush-complete listener — ctx.read() and terminate() were in the same handler with less separation. In 1.2.x, handleLastHttpContent() calls ctx.write() + ctx.read() with no listener, while terminate() fires later in a separate path (HttpServerOperations.operationComplete()). This adds more event loop tasks between ctx.read() and terminate(), widening the window."

### Workaround
I did try setting reactor.netty.http.server.lastFlushWhenNoRead=true as a workaround and it seemed to work from my testing, but as we are not familiar with this property and whether we would be using it correctly, did not proceed to use it long term.

**Comment by violetagg** [comment]
@alexexternaltest35-ctrl Can you try the snapshot version, we might have a fix for this?

**Comment by alexexternaltest35-ctrl** [comment]
@violetagg are you saying it is potentially fixed in latest 1.2 snapshot branch and is there an exact commit we can reference for the fix?

Edit this one: https://github.com/reactor/reactor-netty/commit/3b039229fe4b425ee14cbfd0c647d3abe61441f6?

**Comment by violetagg** [comment]
> [@violetagg](https://github.com/violetagg) are you saying it is potentially fixed in latest 1.2 snapshot branch and is there an exact commit we can reference for the fix?
> 
> Edit this one: [3b03922](https://github.com/reactor/reactor-netty/commit/3b039229fe4b425ee14cbfd0c647d3abe61441f6)?

yes this one

**Comment by alexexternaltest35-ctrl** [comment]
> > [@violetagg](https://github.com/violetagg) are you saying it is potentially fixed in latest 1.2 snapshot branch and is there an exact commit we can reference for the fix?
> > Edit this one: [3b03922](https://github.com/reactor/reactor-netty/commit/3b039229fe4b425ee14cbfd0c647d3abe61441f6)?
> 
> yes this one

Do you know if this issue is also present in 1.3 and if so, is there is a fix for that version as well?

**Comment by alexexternaltest35-ctrl** [comment]
Also, I do not see the fix in Maven as it would be part of 1.2.18, which has not been published AFAIK. Do you have an ETA?

**Comment by violetagg** [comment]
@alexexternaltest35-ctrl Our temporary release plan is here https://github.com/reactor/reactor-netty/milestones

Yes the fix is applied to both versions
